### ISSUE TITLE
transform-class-properties: treatment of undefined properties (i.e. properties without default values)

### ISSUE THREAD

**Issue by mbrowne** [babel/babel#5056]
If you declare an instance property in a class but don't define it, the transform-class-properties plugin does not include any mention of the property at all in the transformed source code.

For example:

	class Demo {
		foo;
	}
	
Currently transpiles to:
	
	Demo = function Demo() {
		_classCallCheck(this, Demo);
	};
	
I propose that it should transpile to this instead:
	
	Demo = function Demo() {
		_classCallCheck(this, Demo);
		this.foo = undefined;
	};

That way, both properties with default values and without would appear in the source code and be enumerable. For example:

	let d = new Demo();
	'foo' in d;  // this would return true  (it currently returns false)

Would it be in violation of the spec somehow to implement this change? If not, I can create a pull request for this. Thanks you.

**Comment by loganfsmyth** [comment]
Sorry it's taken so long to get back to you, I see you've already posted a PR. While I appreciate the motivation, this is currently the behavior listed in the specification. https://tc39.github.io/proposal-class-public-fields/#initialize-public-instance-fields specifically is

<img width="1081" alt="screen shot 2017-01-27 at 7 31 22 pm" src="https://cloud.githubusercontent.com/assets/132260/22393935/42786c44-e4c7-11e6-825d-0e3335c09730.png">

Step 9.a.iii has no `else` behavior, so the current behavior of performing no action is correct.

You'd likely be best to create a new issue in the spec repo. The one you posted in is focused on placing properties on the prototype, which is unrelated to `undefined` behavior.

**Comment by mbrowne** [comment]
Thank you for the clarification. I looked at that spec but wasn't sure if the absence of `else` behavior had to be taken so literally, since the property is still `undefined` after all - now I know. I'll take it up on the spec repo as you suggested.

**Comment by loganfsmyth** [comment]
Yeah, take an example like
```
class Base {
  foo = 4;
}
class Child extends Base {
  foo;
}
```
the value is `4`, if it assigned `undefined` in the general case, it'd override the value.

**Comment by mbrowne** [comment]
Ah, good point. But I think it depends on people's expectations based on programming in other languages with classes...for example, in PHP it'd also override the value:

    class Base {
        public $foo = 4;
    }

    class Child extends Base {
        public $foo;
    }

    $c = new Child();
    var_dump($c->foo);  //null

After thinking about it a bit, I'm leaning toward the opinion that overriding the value is preferable, even if it's initially surprising to some users. If I'm explicitly declaring a property in a subclass, I don't think I'd want the addition of a default value for that property in the parent class to change the behavior of the child class -- especially if the parent class is maintained by a 3rd party.

However, if we did want to prevent overriding, it would be easy enough to do in a concise way:

	function Child() {
	    Base.apply(this, arguments);
	    this.foo = this.foo;
	}

(And there are other more explicit/verbose options as well of course.)

As you suggested, we should probably move this discussion to the spec repo. I'll open a new issue there when I have a moment.

**Comment by mbrowne** [comment]
Created the new issue in the spec repo: https://github.com/tc39/proposal-class-public-fields/issues/57

**Comment by mbrowne** [comment]
Hi, not sure if you've been following along at the link above, but it turns out this was already updated in the spec repo, it's just that the github page wasn't updated. So I believe my pull request is now officially ready to go.

**Comment by loganfsmyth** [comment]
Cool, thanks for looking into it more.

Since this is a breaking change, we'll need to put it behind an opt-in flag on the plugin for Babel 6 or else it could break people's code. We can certainly enable this behavior by default for Babel 7 though.

**Comment by mbrowne** [comment]
Sounds good, thanks. Yes, I suppose it would break people's code if they were redeclaring properties like in your inheritance example above. Or if they were counting on uninitialized properties being non-enumerable for some strange reason.

**Comment by mbrowne** [comment]
I can help out with putting it behind a flag for Babel 6 if anyone is interested. But for starters, can this be merged to the Babel 7 branch?

**Comment by littledan** [comment]
Class property declarations seem likely to get to Stage 3 in a TC39 meeting coming soon. When they are moved to the stage 3 preset, is there any way that the semantics in that preset could somehow switch to the new semantics? It would be a shame if a larger group of people started using class property declarations at this point, thinking that they were stable, where they are actually going to change semantics.

**Comment by mbrowne** [comment]
@littledan What would you regard as the big changes? Also, it seems that there could be more significant changes in the future, especially if there's some agreement about creating properties on the prototype and not only within the constructor. (Aside: I agree with Adam Miller's comment [here](https://github.com/tc39/proposal-class-public-fields/issues/38#issuecomment-238451020.)

**Comment by littledan** [comment]
@mbrowne I think this semantic mismatch is a pretty major change. We'll see this week if it's the last one, but there's been consensus for a number of months that we want to treat missing initializers in this way.

@ljharb I assume closing this issue was a mistake, right?

**Comment by ljharb** [comment]
Yes, that was just me rebasing test262 master and pushing to my fork. Thanks for the reopen

**Comment by hzoo** [comment]
I'm going to report ^ to github since it seems like a bug.

Also the `this.foo = undefined` change is in Babel already but under the spec flag. https://github.com/babel/babel/blob/7.0/packages/babel-plugin-transform-class-properties/src/index.js#L36 https://github.com/babel/babel/pull/4544?

so in the config can use ` ["transform-class-properties", { "spec": true }]` instead.

**Comment by littledan** [comment]
@hzoo Oh thanks for explaining, I didn't realize. Will this flag be on by default in Babel 7?

**Comment by hzoo** [comment]
I'm not sure yet but we could - it also uses Object.defineProperty instead of assignment

**Comment by mbrowne** [comment]
> Also the this.foo = undefined change is in Babel already but under the spec flag. 

Great, thank you. The correct behavior would be for this to be on by default. I think that to keep the old behavior by default (which the spec authors may never even have intended in the first place - I think it was a mistake in the spec description) would be misleading. After that's done then this issue could be closed as far as I'm concerned.

Thanks all!

**Comment by mbrowne** [comment]
...Just to be clear, I'm specifically referring to setting uninitialized properties to `undefined`, which is how it should behave in Babel 7 regardless of flags (well, unless someone wanted to add a `legacy` flag to mimic Babel 6 behavior but I don't see any reason for that). As to usage of `Object.defineProperty` or anything else that's different with the `spec` flag, I don't have an opinion on that.

**Comment by hzoo** [comment]
Oh ok ^ yeah we can make that change too @peey if you want to add this in (or someone else)

**Comment by littledan** [comment]
I wouldn't be so sure it was a mistake; the spec's been through a lot of iterations. `Object.defineProperty` is the current semantics, but it used to be based on `=`.

BTW, minor point: does it really expand into `undefined` literally? Would `void 0` be a better choice, in the unlikely event that `undefined` is shadowed?

**Comment by hzoo** [comment]
The void 0/undefined issue is generic so if it's doing undefined currently we can output void 0 (this came up before), so that's fine/easy to fix if that's the case

**Comment by hzoo** [comment]
Ok I think we will make the default behavior the current spec (defineProperty) and then make current behavior spec in v7 then.

**Comment by mbrowne** [comment]
@hzoo Just to be clear, this means that the default behavior will be to set uninitialized properties to `undefined`, right? (Or `void 0` if deemed necessary.)

@littledan, BTW, my comment about the spec authors' intent was specifically in reference to initializing to `undefined`, not `Object.defineProperty` (based on [this comment](https://github.com/tc39/proposal-class-public-fields/issues/57#issuecomment-275884665)).

**Comment by mbrowne** [comment]
@littledan Oh, haha, just realized that you're the same person that that comment refers to. Well I guess the history doesn't matter so much now; the important thing is to honor the current spec (especially with something like the `undefined` issue, IMO).

**Comment by kedromelon** [comment]
just came across this -- I believe this (and its PR: https://github.com/babel/babel/pull/5115) can be closed now that https://github.com/babel/babel/pull/6076 has been merged?

**Comment by hzoo** [comment]
Yeah, Fixed by #6076
